7801 add more by-dnode routines

ahrens opened the pull request:
7801 add more by-dnode routines

    Reviewed by: Matthew Ahrens <mahrens@delphix.com>
    Reviewed by: Brian Behlendorf <behlendorf1@llnl.gov>

    Add *_by_dnode() routines for accessing objects given their
    dnode_t *, this is more efficient than accessing the object by
    (objset_t *, uint64_t object).  This change converts some but
    not all of the existing consumers.  As performance-sensitive
    code paths are discovered they should be converted to use
    these routines.

Ported from: https://github.com/zfsonlinux/zfs/commit/0eef1bde31d67091d3deed23fe2394f5a8bf2276

ahrens commented:
@zettabot go

prakashsurya commented:
this failed lint:
```
"../../common/fs/zfs/dmu_tx.c", line 101: warning: argument unused in function: os in dmu_tx_hold_dnode_impl (E_FUNC_ARG_UNUSED)
"/var/jenkins/fsroot/workspace/02-checkout/openzfs/usr/src/uts/common/fs/zfs/dmu_tx.c", line 107: warning: function returns value which is sometimes ignored: refcount_add (E_FUNC_RET_MAYBE_IGNORED2)
```

ahrens commented:
@zettabot go

ahrens commented:
@grwilson @pzakha could you take a look at this?

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

pzakha commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -1149,8 +1145,39 @@ zap_add(objset_t *os, uint64_t zapobj, const char *key,
 	zap_name_free(zn);
-	if (zap != NULL)	/* may be NULL if fzap_add() failed */
-		zap_unlockdir(zap, FTAG);
+	zap_unlockdir(zap, tag);
```
Doesn't this comment still apply?
```/* may be NULL if fzap_add() failed */```

ahrens replied:
Yes I think so, I'll revert this.

pzakha commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -1299,6 +1320,34 @@ zap_remove_norm(objset_t *os, uint64_t zapobj, const char *name,
+	err = zap_lockdir_by_dnode(dn, tx, RW_WRITER, TRUE, FALSE, FTAG, &zap);
+	if (err)
+		return (err);
+	err = zap_remove_impl(zap, name, MT_EXACT, tx);
```
Will need to be rebased on https://github.com/openzfs/openzfs/commit/8f1750dd675844e8b2de749a6613c07b8e541711

pzakha reviewed (commented):
Logic looks good, except for the comment I've added about ```/* may be NULL if fzap_add() failed */```

I've also added some code style suggestions but you can probably ignore them since we are porting from ZoL.

ahrens commented:
@zettabot go

ahrens commented:
@pzakha I've made all the changes you suggested, and I'll try to get them ported to ZoL too.

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

pzakha reviewed (approved):
There's still a few
```
if (err)
```
```
if (txh)
```
But otherwise functionally it looks good.

Thanks Matt

bzzz77 committed:
7801 add more by-dnode routines

Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>

Add *_by_dnode() routines for accessing objects given their
dnode_t *, this is more efficient than accessing the object by
(objset_t *, uint64_t object).  This change converts some but
not all of the existing consumers.  As performance-sensitive
code paths are discovered they should be converted to use
these routines.

Ported from: https://github.com/zfsonlinux/zfs/commit/0eef1bde31d67091d3deed23fe2394f5a8bf2276

Closes #276

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go

ahrens commented:
@zettabot go
